### ISSUE TITLE
Is it possible to access the previous row?

### ISSUE THREAD

**Issue by jfly** [hunterae/table-for#8]
When rendering columns, it would be nice to have some way of accessing the entire array we're rendering, and our current index inside of it. This would let us access the previous row.

**Comment by hunterae** [comment]
Hey @jfly . I agree completely. For the current index to be available, I need to make a change to the underlying "blocks" gem, but that wouldn't be terribly difficult and it would definitely be a nice feature. 

I am in the process of redoing the "blocks" gem from scratch to make it more powerful, but should be able to squeeze that index feature into the current edition fairly easily.

In the meantime, here's a possibly workaround. The collection of records being rendered in the table is always accessible to you. Using that collection, you could fetch the index of the current record. What follows is four different ways you could do so (Please note: your table.column call will only need to take 3 parameters as below if you go with the second approach. Additionally, these solutions will output index 0 for the first row. It is uncertain whether we'd want to do a zero-based index for your proposed feature. Perhaps that could be configurable):

```
<%= table_for @users do |table| %>
  <% table.column :index do |user, column, column_options| %>
    Start of row <%= @users.index(user) %><br>
    OR start of row <%= column_options[:records].index(user) %><br>
    OR start of row <%= table.init_options[:records].index(user) %><br>
    OR start of row <%= table.global_options[:records].index(user) %><br>
  <% end %>
<% end %>
```

Hope this helps for the mean time.

**Comment by jfly** [comment]
Thanks for the quick response! Doing a `.index(` for each row feels little too `O(n)` to me, so we'll probably keep a running count ourselves for now. Do you know if there are there any downsides to doing so?

It would definitely be be really nice to have access to the (0 indexed, IMO) index!

**Comment by hunterae** [comment]
Although it would add a extra call, rendering a table is already ```O(n)``` complexity, so it would really depend on whether ```.index``` is a complicated / time-consuming operation or not since you'd be calling it n times.

I can definitely add the current_index in, but it might take a few days.

For a little clarification / background, it has to do with this declaration declared here: https://github.com/hunterae/table-for/blob/master/app/views/table_for/_table_for.html.erb#L38-L43:

```
<% table.define :data_row do |record, options| %>
  <% table.set_current_record(record) %>
  <%= content_tag :tr, table.call_each_hash_value_with_params(options[:data_row_html], record) do %>
    <%= table.render :data_column, record, :collection => table.columns %>
  <% end %>
<% end %>
```

More particularly, it involves the line (which renders a block for each item in the collection):
```
<%= table.render :data_column, record, :collection => table.columns %>
```

Which gets handled by the blocks gem here: https://github.com/hunterae/blocks/blob/master/lib/blocks/base.rb#L163-L183

So I suppose it would involve changing this line: https://github.com/hunterae/blocks/blob/master/lib/blocks/base.rb#L168
to use each_with_index.

Since there's still several moving parts and corresponding tests, it would be an instant fix, but likely by mid-end of week.

**Comment by jfly** [comment]
On further thought, we're actually trying to access the previous row inside of a `data_column_html: { class: lambda do |record, column| ... ` callback, not when rendering a column. Both would be super useful, though!

**Comment by hunterae** [comment]
Ah, gotcha. In that case, you could do something like:

```:data_column_html => { class: lambda { |record| "previous-index-#{@users.index(record)-1}" } }```

but definitely, just having access to the current index would be helpful.

**Comment by hunterae** [comment]
@jfly : this feature is now implemented in version 3.5.0. You can access the current row number with table.current_index or table.current_row_index.

Example usage:

```
<%= table_for @users do |table| %>
    <% table.column :id do |user| %>
      "User id #{user.id} rendering with row index #{table.current_index}"
    <% end %>
<% end %>
```

In my example from my previous comment, might now look something like:

```
:data_column_html => { class: lambda { |record| "previous-index-#{table.current_index-1}" } }
```
